include/deal.II/fe/fe_values_coupling.h

```diff
@@ -0,0 +1,557 @@
+#ifndef DOXYGEN
+template <int dim, int spacedim>
+class FEValuesBase;
+#endif
+
+namespace FEValuesViews
+{
+  /**
+   * A class that provides a renumbered view to a given FEValuesViews object.
+   *
+   * In general, the order of the degrees of freedom and quadrature points
+   * follows the one of the FEValues object. However, in some cases, it is
+   * convenient to group together degrees of freedom and quadrature points in a
+   * different order. This class provides a view to a given FEValuesViews
+   * object, where the degrees of freedom and quadrature points are renumbered
+   * according to the given renumbering vectors.
+   */
+  template <typename ViewType>
+  class ReorderedView : public Subscriptor
+  {
+  public:
+    /**
+     * An alias for the data type of values of the underlying view.
+     */
+    using value_type = typename ViewType::value_type;
+
+    /**
+     * An alias for the data type of gradients of the underlying view.
+     */
+    using gradient_type = typename ViewType::gradient_type;
+
+    /**
+     * An alias for the data type of the product of a @p Number and the values
+     * of the underlying view type.
+     */
+    template <typename Number>
+    using solution_value_type = typename ProductType<Number, value_type>::type;
+
+    /**
+     * An alias for the data type of the product of a @p Number and the gradients
+     * of the underlying view type.
+     */
+    template <typename Number>
+    using solution_gradient_type =
+      typename ProductType<Number, gradient_type>::type;
+
+    /**
+     * Construct a new ReorderedView object.
+     *
+     * The @p dof_renumbering vector is used to renumber the degrees of freedom,
+     * while the @p quadrature_renumbering vector is used to renumber the
+     * quadrature points. An empty renumbering vector simply means that no
+     * renumbering is performed.
+     *
+     * @note The renumbering vectors may *not* match the dimensions of the
+     * underlying view, i.e., you could use this class to only run over half of
+     * the degrees of freedom, and integrating only over half of the current
+     * cell, if you wish to do so. The important part is that every entry of
+     * each renumbering vector is a legal value for the underlying view object.
+     * We allow the dof renumbering vector to contain
+     * `numbers::invalid_unsigned_int` values, which are ignored, and produce
+     * zero values, gradients, hessians, etc. for the corresponding shape
+     * function.
+     *
+     * @param view The underlying FEValuesViews object.
+     * @param dof_renumbering A renumbering vector for the degrees of freedom.
+     * @param quadrature_renumbering A renumbering vector for the quadrature
+     * points.
+     */
+    ReorderedView(const ViewType &                 view,
+                  const std::vector<unsigned int> &dof_renumbering        = {},
+                  const std::vector<unsigned int> &quadrature_renumbering = {});
```

Comment:
Can you document somehow the direction in which the renumbering works? I think this can be inferred from the comment that one can select a subset of indices, but it might be useful to give an example.

Reply:
ok.